### ISSUE TITLE
MMEX 1.6.4  Import from CSV error. PAYEE_V1: Exception constraint failed[2067]

### ISSUE THREAD

**Issue by B-What** [moneymanagerex/moneymanagerex#6057]
MMEX version:
 - [ ] 1.6.0
 - [ ] 1.6.1
 - [ ] 1.6.2
 - [ ] 1.6.3 Beta
 - [x] 1.6.4 64bit

Note: bug reporters are expected to have verified the bug still exists
either in the last stable version of MMEX or on updated development code
(master branch).

Operating System:
 - [x] Windows 64bit
 - [ ] Mac OSX
 - [ ] Linux 

# Description of the bug

When importing from CSV, the result is successful, upon completion an error message pops up (see screenshot) PAYEE_V1: Exception constraint failed[2067]: UNIQUE constraint failed: PAYEE_V1.PAYEENAME. The import is successful, although the payee and category fields are imported as "unknown".

I tried to revert to MMEX version 1.6.3, however the database can no longer be opened by version 1.6.3.

![image](https://github.com/moneymanagerex/moneymanagerex/assets/137909977/64cd8cd4-2697-4e72-8f82-dcb574df2e1b)

![image](https://github.com/moneymanagerex/moneymanagerex/assets/137909977/2187e31c-4086-426a-9d30-c65e0bb34f9a)

# Reproduction

Is the bug reproducible? 
 - [x] Always 
 - [ ] Randomly 
 - [ ] Happened only once

Reproduction steps:  
Import from CSV > set Date, Amount, Payee as always > Click Import > Import dialog box says import was successful > Click "OK" > MoneyManagerEx Error dialog box pops up with the following error message " PAYEE_V1: Exception constraint failed[2067]: UNIQUE constraint failed: PAYEE_V1.PAYEENAME "

Expected result:
Import from CSV into database, with Date, Amount, Payee all showing up in the correct fields.

Actual result:
The CSV is imported, Date and Amount are correct, but all Payee's are "Unknown" and all Categories are "Unknown".

# Additional information

16:49:04: PAYEE_V1: Exception constraint failed[2067]: UNIQUE constraint failed: PAYEE_V1.PAYEENAME, {
    "PAYEEID": -1,
    "PAYEENAME": "EMPLOYEE WAGES",
    "CATEGID": -1,
    "NUMBER": "",
    "WEBSITE": "",
    "NOTES": "",
    "ACTIVE": 1,
    "PATTERN": ""
}

**Comment by B-What** [comment]
On the preview pane, switching to the "Payee" tab shows "missing" under Status. Not sure if this is related?

![image](https://github.com/moneymanagerex/moneymanagerex/assets/137909977/561f15d2-6987-447d-932e-8f8dcdb53dd1)

**Comment by kevyndev** [comment]
I can't reproduce this running the last 1.6.4, but this was already reported in #5947 while running 1.6.4 beta.

can you confirm what version are you using by going "Help > About"

**Comment by tactilis** [comment]
@B-What
> I tried to revert to MMEX version 1.6.3, however the database can no longer be opened by version 1.6.3.

When you opened your .mmb file in 1.6.4 for the first time, the format of the database was upgraded from v17 to v18.

MMEX 1.6.4 will have created a file named `<your_database_name>.mmb_upgrade_v17_YYYY-MM-DD.bak`. This is a backup copy of your v17 format database that you were using in 1.6.3.

If you want to revert to 1.6.3 until the problem you are having is solved, then rename the above .bak file to <your_database_name>.mmb and open it in 1.6.3.

That file will, of course, not contain any transactions that you entered using 1.6.4 but it will allow you to continue to use MMEX until a fix is found.

**Comment by n-stein** [comment]
> On the preview pane, switching to the "Payee" tab shows "missing" under Status. Not sure if this is related?

This just means a payee with that name is not already found in your database and a new one will be created.

I think the issue here is due to case differences. Duplicating the same payee with multiple case styles results in this issue.

**Comment by B-What** [comment]
Thanks for the feedback guys.

About:
![16879404810337779448904577821876](https://github.com/moneymanagerex/moneymanagerex/assets/137909977/719b6078-ccee-441d-b521-5ef7ae1b9412)


I found a fairly quick workaround by converting my .CSV to .QIF and using the .QIF importer, which imported all the transactions perfectly and even auto-assigned categories to the known payees as expected.

I'll continue to use this workaround for now, but am happy to help provide feedback and test wherever necessary to help resolve this bug.

**Comment by tactilis** [comment]
> I think the issue here is due to case differences. Duplicating the same payee with multiple case styles results in this issue.

Yes.

It should not be possible to have multiple Payees in the database with the same name except for the case of letters.

Creation of such duplicates is prevented now - but at some point in the past, due to a bug, it was allowed.

@B-What
1. Please go to `Tools | Organise Payees` and look for **_any_** duplicate Payee names, differing only in letter case. Check particularly 'REHMAN UNISE...', which is the name that failed during your CSV import.

2. Rename one of them (eg. add ZZZ to the name).

3. Try your CSV import. Does it work now?

4. Now you can use `Tools | Relocation of | Payees` to relocate 'Rehman Unise...ZZZ' to  'REHMAN UNISE...'

**Comment by B-What** [comment]
@tactilis 

I have checked the Payees. There are no duplicates where the only difference is CASE.

Although, I do have many "REHMAN UNISE..." due to the fact that every time I import, I import from a bank statement. The bank statement lists the transaction including the card number and transaction date. Basically this means every transaction done with the card is a "unique" transaction as, at the very least, the date is not the same. 

For this reason I have thousands of payees in my payee database, even though it is simply the same Payee, the only difference is that the transaction was imported with a different date in the text. 

IE 
~ REHMAN UNISEX SALON 3005**475   01 JUN
vs
~ REHMAN UNISEX SALON 3005**475   15 JUN

I have tried to rectify this by using excel to truncate my CSVs  to only the first 20 Characters, before importing them, that way the date is excluded from the end of the text string, (but because this is not a perfect science, it sometimes also excludes important data from the text string) ,  MMEX then sees the truncated payee as the same payee as past imports, and works as expected.

If I use excel to truncate to 20 Characters, I get:
~ REHMAN UNISEX SALON
vs
~ REHMAN UNISEX SALON 3005**475   15 JUN


The issue here is that on a different transaction I might lose data from the END of the text string
 ie:
~ MONEY TRANSFER TO EMPLOYEE #123
Gets truncated to
~ MONEY TRANSFER TO EM

So the workaround is not really a workaround, and I have just allowed my Payee database to grow with payees. The negative to this is of course the auto-assigning of the categories does not work.

**Comment by tactilis** [comment]
> Although, I do have many "REHMAN UNISE..." due to the fact that every time I import, I import from a bank statement. The bank statement lists the transaction including the card number and transaction date. Basically this means every transaction done with the card is a "unique" transaction as, at the very least, the date is not the same.
>
> For this reason I have thousands of payees in my payee database, even though it is simply the same Payee, the only difference is that the transaction was imported with a different date in the text.

This is precisely what the new feature of `Pattern match payees` in 1.6.4 is designed to deal with. For a Payee e.g. REHMAN UNISEX SALON you can set up one or more wildcard patterns to match the bank's CSV text to your single payee name.

Define the pattern(s) here:

![image](https://github.com/moneymanagerex/moneymanagerex/assets/4060092/1c47c2c5-f3d7-4c41-9bf8-2623f77c2239)

You can go directly to the the above screen from the Payee tab of the Import dialog. Double click on the payee or right-click and select Organise Payees.

If you are interested, you can read about the development of Pattern Match Payees here https://github.com/moneymanagerex/moneymanagerex/issues/3148.

It's a really slick facility if set up and used properly.

Using it won't fix your existing multiple payees but you can avoid generating more in the future and it will avoid you having to truncate payee names in Excel.

**Comment by B-What** [comment]
Brilliant feature, thank you for the advice @tactilis

**Comment by n-stein** [comment]
Fixed two bugs:

1. Payee and Category creation was case insensitive, but the vector search for payees/categories was not. So if a payee didn't previously exist and multiple cases for the new payee were in the file the system would attempt to create the same payee once for each different case.
2. Importing categories did not work properly if the import file had nested categories and there were spaces on either side of the `:` delimiter. This means round-trip export/import did not function if either of the category delimiters `  :  ` or `: ` were set in options.

**Comment by tactilis** [comment]
@n-stein
> Importing categories did not work properly if the import file had nested categories and there were spaces on either side of the : delimiter. This means round-trip export/import did not function if either of the category delimiters  : or : were set in options.

You write about _either_ of the delimiters ` : ` or `: ` being set in options. Can I check that you are aware that the three delimiters offered in the dropdown are not the only possibilities for the category delimiter.

It is possible for the user to type any number of characters (with or without spaces) in the field:

![image](https://github.com/moneymanagerex/moneymanagerex/assets/4060092/a4997ca3-d146-41a9-bdb9-d7ddb9937d72)

Does this affect importing?

**Comment by n-stein** [comment]
>Does this affect importing?

Yes. Since nested categories were introduced the `:` character has been used to delimit strings on import. If you attempted to upload `BillsxElectricity` MMEX would simply create a new category called BillsxElectricity rather than assigning a transaction to the `Electricity` category under parent category `Bills`. 

I think import needs to be restricted to the `:` delimiter since this is the only reserved character in category names. If you used another character such as `x` you wouldn't be able to have a category with `x` in its name like `Taxes` or MMEX would incorrectly interpret the category name as a `parent : child` relationship `Ta : es`.

At the same time I realize that round-trip export/import should be preserved. I think this means we should also restrict export to replace the category delimiter with `:` as well.

**Comment by tactilis** [comment]
@n-stein 
> At the same time I realize that round-trip export/import should be preserved. I think this means we should also restrict export to replace the category delimiter with : as well.

Yes. Exactly that.


<br>

> I think import needs to be restricted to the : delimiter since this is the only reserved character in category names. If you used another character such as x you wouldn't be able to have a category with x in its name like Taxes or MMEX would incorrectly interpret the category name as a parent : child relationship Ta : es.

To be clear, I was not proposing that you need to accommodate any delimiter for export & import other than `:`. I was just checking that you were aware that a delimiter could be typed in the Options.

<br>

In truth, I think the whole `:` / `: ` / ` : ` / `Enter any delimiter text you like` choice was a bit of a UI misstep because it has a complicated the code in several places. Looking back, it arose in response to one user complaining that the `category:subcategory` pop-ups on the pie-charts were unreadable on high dpi monitors, so they wanted a space after the colon.

![image](https://github.com/moneymanagerex/moneymanagerex/assets/4060092/e29a1d3b-2fbf-4b81-8043-86ee18987f9c)

<br>

- Solving the pop-up readability problem would have been best.

- Adding choices of  `:` / `: ` / ` : ` probably seemed like the easiest option to satisfy everyone. But I would have argued strongly for no choices (so everyone got a display of `: ` ) might have been easier.

- Allowing users to enter their own delimiter seems unnecessary.

Sometimes providing no choice is the best way forward (IMO). But I guess we are stuck with the current implementation now.

<br>

**Comment by n-stein** [comment]
Exporting is now set to the standard delimiter `:`

**Comment by B-What** [comment]
Hi guys, just to update from my side. I still have the same issue importing csv, and in many cases It updates my Payee to unknown as well as the category. I have now switched to importing via Qif as this does not throw the above mentioned exception, and I can proceed just fine like this, but obviously please note that the issue still exists via csv import.
I tried to pattern match with the qif imports and it did not work. I know this because I then deleted the import and tried again without pattern match and got the exact same auto assignments on categories. IE selecting the "pattern match" tick box at import made no difference to the number of transactions that were automatically assigned a category.

Kind regards,

**Comment by tactilis** [comment]
@B-What
> Hi guys, just to update from my side. I still have the same issue importing csv

In what version of MMEX?  The fix will be in MMEX 1.7.0 (coming soon)

**Comment by n-stein** [comment]
Closing in favor of #7090 which is reported on the current version of MMEX. If this issue still exists, please update there or create a new issue.
